be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/SocialMediaHandler.scala

```diff
@@ -37,49 +41,38 @@ class SocialMediaHandler(dbRef: => AskableActorRef) extends MessageHandler {
+    val ask =
+      for {
+        _ <- extractLaoChannel(rpcMessage, "Server failed to extract LAO id for the broadcast")
+        _ <- extractParameters[AddChirp](rpcMessage, "Server failed to extract chirp id for the broadcast")
+        _ <- dbActor ? DbActor.WriteAndPropagate(rpcMessage.getParamsChannel, rpcMessage.getParamsMessage.get)
+      } yield ()
+
+    Await.ready(ask, duration).value match {
+      case Some(Success(_)) =>
+        val (chirp_id, channelChirp, data, broadcastChannel) = parametersToBroadcast[AddChirp](rpcMessage)
+        val notifyAddChirp: NotifyAddChirp = NotifyAddChirp(chirp_id, channelChirp, data.timestamp)
+        Await.result(dbBroadcast(rpcMessage, channelChirp, Base64Data.encode(notifyAddChirp.toJson.toString), broadcastChannel), duration)
+      case Some(Failure(ex: DbActorNAckException)) => Right(PipelineError(ex.code, s"handleAddChirp failed : ${ex.message}", rpcMessage.getId))
+      case _                                       => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, unknownAnswerDatabase, rpcMessage.getId))
     }
   }
 
   def handleDeleteChirp(rpcMessage: JsonRpcRequest): GraphMessage = {
-    writeAndPropagate(rpcMessage) match {
-      case Left(_) =>
-        val channelChirp: Channel = rpcMessage.getParamsChannel
-        channelChirp.decodeChannelLaoId match {
-          case Some(lao_id) =>
-            val broadcastChannel: Channel = generateSocialChannel(lao_id)
-            rpcMessage.getParamsMessage match {
-              case Some(params) =>
-                val chirp_id: Hash = params.message_id
-                val timestamp: Timestamp = params.decodedData.get.asInstanceOf[DeleteChirp].timestamp
-                val notifyDeleteChirp: NotifyDeleteChirp = NotifyDeleteChirp(chirp_id, channelChirp, timestamp)
-                val broadcastData: Base64Data = Base64Data.encode(notifyDeleteChirp.toJson.toString)
-                Await.result(dbBroadcast(rpcMessage, rpcMessage.getParamsChannel, broadcastData, broadcastChannel), duration)
-              case None => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, "Server failed to extract chirp id for the broadcast", rpcMessage.id))
-            }
-          case None => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, "Server failed to extract LAO id for the broadcast", rpcMessage.id))
-        }
-      case error @ Right(_) => error
-      case _                => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, unknownAnswerDatabase, rpcMessage.id))
+    val ask =
+      for {
+        _ <- extractLaoChannel(rpcMessage, "Server failed to extract LAO id for the broadcast")
+        _ <- extractParameters(rpcMessage, "Server failed to extract chirp id for the broadcast")
+        _ <- dbActor ? DbActor.WriteAndPropagate(rpcMessage.getParamsChannel, rpcMessage.getParamsMessage.get)
+      } yield ()
+
+    Await.ready(ask, duration).value match {
```

Comment:
Even though we still have one chained Await, it's way better and easily readable than before!